[FrameworkBundle] Deprecate not setting some options (uid, validation)

Jean-Beru opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes 
| Tickets       | Fix #51097
| License       | MIT
| Doc PR        | 

This PR deprecates not setting 4 options in framework configuration to match the v7.0 recipe:
* `uid.default_uuid_version` which is `6` by default and is set to `7` by recipe
* `uid.time_based_uuid_version`  which is `6` by default and is set to `7` by recipe
* `validation.email_validation_mode` which is `null` by default and is set to `html5` by recipe

See #51097.

Jean-Beru committed:
[FrameworkBundle] Deprecate not setting some options

Jean-Beru committed:
[FrameworkBundle] Deprecate not setting some options (uid, validation)

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -2316,14 +2325,31 @@ private function addRateLimiterSection(ArrayNodeDefinition $rootNode, callable $
+                        }
+
+                        if (!array_key_exists('time_based_uuid_version', $v['uid'])) {
+                            trigger_deprecation('symfony/framework-bundle', '6.1', 'Not setting the "framework.uid.time_based_uuid_version" config option is deprecated. It will default to "7" in 7.0.');
```
this cannot be a deprecation introduced in FrameworkBundle 6.1 as that version is already released

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php:
```diff
@@ -418,6 +418,15 @@ private function addProvidersSection(ArrayNodeDefinition $rootNode): void
+            ->validate()
+                ->always(function (array $v): array {
+                    if (empty($v['password_hashers'])) {
+                        trigger_deprecation('symfony/framework-bundle', '6.4', 'Not setting the "security.password_hashers" config option is deprecated. It will default to "Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: auto" in 7.0.');
```
This is wrong. This deprecation does not happen in the èsymfony/framework-bundle` package.

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php:
```diff
@@ -418,6 +418,15 @@ private function addProvidersSection(ArrayNodeDefinition $rootNode): void
+            ->validate()
+                ->always(function (array $v): array {
+                    if (empty($v['password_hashers'])) {
+                        trigger_deprecation('symfony/framework-bundle', '6.4', 'Not setting the "security.password_hashers" config option is deprecated. It will default to "Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: auto" in 7.0.');
```
I'm not sure about this deprecation. Default values for a prototyped array node are not merged with provided config. So as soon as any config file (or config prepended by a bundle) adds some other config for password hashers, the config for PasswordAuthenticatedUserInterface would be lost.

I'd suggest keeping that one in the recipe.

nicolas-grekas replied:
They would be merged if we didn't call `performNoDeepMerging`, isn't it?
Why don't we perform deep merging BTW?

nicolas-grekas replied:
Anyway, let's drop this, it's not needed to deprecate not setting this

wouterj replied:
:shrug: a5cfc2207c09f6182333b3b812673202701c33f3 "some bug fixes in DIC config classes"

stof replied:
@nicolas-grekas no. deep merging is about the various configs received as input, not the defaults

nicolas-grekas reviewed (dismissed):
I addressed @stof's review.

nicolas-grekas committed:
[Serializer] Fix deps

nicolas-grekas committed:
feature #51325 [FrameworkBundle] Deprecate not setting some options (Jean-Beru)

This PR was merged into the 6.4 branch.

Discussion
----------

[FrameworkBundle] Deprecate not setting some options

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Fix #51097
| License       | MIT
| Doc PR        |

This PR deprecates not setting 5 options in framework configuration to match the v7.0 recipe:
* `handle_all_throwables` which is `false` by default and is set to `true` by recipe
* `php_errors.log` which use `$debug` value by default and is set to `true` by recipe
* `session.cookie_secure` (if session is enabled) which is `null` by default and is set to `auto` by recipe
* `session.cookie_samesite` (if session is enabled) which is `null` by default and is set to `lax` by recipe
* `session.handler_id` (if session is enabled) which is `session.handler.native_file` by default and is set to `null` by recipe (could be `session.handler.native_file` if `framework.session.save_path` is set)

See #51097.

Commits
-------

7f9998e953 [FrameworkBundle] Deprecate not setting some options

nicolas-grekas commented:
Thank you @Jean-Beru.
